pkg/auth/providers/saml/saml_client.go

```diff
@@ -0,0 +1,385 @@
+
+	provider := SamlProviders[PingName]
+	if provider == nil {
+		return Root
+	}
+
+	storedSamlConfig, err := provider.getSamlConfig()
+	if err != nil {
+		log.Errorf("SAML(PingHandlers): error in getting config: %v", err)
+		return Root
+	}
+	if !storedSamlConfig.Enabled {
+		return Root
+	}
+	InitializeSamlServiceProvider(storedSamlConfig, PingName)
+	return Root
+}
+
+func (s *Provider) getSamlPrincipals(config *v3.SamlConfig, samlData map[string][]string) (v3.Principal, []v3.Principal) {
+	var userPrincipal v3.Principal
+	var groupPrincipals []v3.Principal
+	uid, ok := samlData[config.UIDField]
+	if ok {
+		userPrincipal = v3.Principal{
+			ObjectMeta:    metav1.ObjectMeta{Name: s.UserType + "://" + uid[0]},
+			Provider:      s.Name,
+			PrincipalType: s.UserType,
+			Me:            true,
+		}
+
+		displayName, ok := samlData[config.DisplayNameField]
+		if ok {
+			userPrincipal.DisplayName = displayName[0]
+		}
+
+		userName, ok := samlData[config.UserNameField]
+		if ok {
+			userPrincipal.LoginName = userName[0]
+		}
+
+		groups, ok := samlData[config.GroupsField]
+		if ok {
+			for _, group := range groups {
+				group := v3.Principal{
+					ObjectMeta:    metav1.ObjectMeta{Name: s.GroupType + "://" + group},
+					DisplayName:   group,
+					Provider:      s.Name,
+					PrincipalType: s.GroupType,
+					MemberOf:      true,
+				}
+				groupPrincipals = append(groupPrincipals, group)
+			}
+		}
+	}
+	return userPrincipal, groupPrincipals
+}
+
+// HandleSamlAssertion processes/handles the assertion obtained by the POST to /saml/acs from IdP
+func (s *Provider) HandleSamlAssertion(w http.ResponseWriter, r *http.Request, assertion *saml.Assertion) {
+	var groupPrincipals []v3.Principal
+	var userPrincipal v3.Principal
+
+	if relayState := r.Form.Get("RelayState"); relayState != "" {
+		// delete the cookie
+		s.ClientState.DeleteState(w, r, relayState)
+	}
+
+	samlData := make(map[string][]string)
+
+	for _, attributeStatement := range assertion.AttributeStatements {
+		for _, attr := range attributeStatement.Attributes {
+			attrName := attr.FriendlyName
+			if attrName == "" {
+				attrName = attr.Name
+			}
+			for _, value := range attr.Values {
+				samlData[attrName] = append(samlData[attrName], value.Value)
+			}
+		}
+	}
+
+	config, err := s.getSamlConfig()
+	if err != nil {
+		writeError(w, 500, "SAML: Error getting saml config %v", err, "Server error while authenticating")
+		return
+	}
+
+	userPrincipal, groupPrincipals = s.getSamlPrincipals(config, samlData)
+	allowedPrincipals := config.AllowedPrincipalIDs
+
+	allowed, err := s.UserMGR.CheckAccess(config.AccessMode, allowedPrincipals, userPrincipal, groupPrincipals)
+	if err != nil {
+		writeError(w, 500, "SAML: Error during login while checking access %v", err, "Server error while authenticating")
+		return
+	}
+	if !allowed {
+		writeError(w, 403, "SAML: User does not have access %v", err, "User does not have access")
+		return
+	}
+
+	if !config.Enabled {
```

Comment:
as discussed, make this decision based off of claims in the jwt, not the state of enabled

Reply:
when you change this logic, please reorganize to make it very clear that one code path is for the "testing" scenario and the other is for normal login. Like, even make different functions for each of those code paths for this function to call